Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm

```diff
@@ -1333,4 +1333,92 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                fontURLs.append(fontURL);
+                RELEASE_LOG(Process, "Registering font url %s", fontURL.string().utf8().data());
+            }
+        }
+
+        RunLoop::protectedMain()->dispatch([weakThis, fontURLs = WTFMove(fontURLs)] {
+            if (RefPtr protectedThis = weakThis.get()) {
+                for (auto& process : protectedThis->m_processes)
+                    process->send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(fontURLs)), 0);
+                protectedThis->m_additionalFontURLs = WTFMove(fontURLs);
+            }
+        });
+    });
+
+    if (m_additionalFontURLs) {
+        process.send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts(*m_additionalFontURLs)), 0);
+        return;
+    }
+    static bool hasStartedCachingUserInstalledFonts = false;
+    if (!hasStartedCachingUserInstalledFonts) {
+        dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), blockPtr.get());
+        hasStartedCachingUserInstalledFonts = true;
+    }
+}
+
+void WebProcessPool::registerAssetFont(ASCIILiteral fontName, WebProcessProxy& process)
+{
+    auto nsFontName = fontName.createNSString();
+    NSDictionary *attributes = @{ (NSString *)kCTFontFamilyNameAttribute: nsFontName.get(), (NSString *)kCTFontRegistrationScopeAttribute: @(kCTFontPriorityComputer) };
+    auto description = adoptCF(CTFontDescriptorCreateWithAttributes((__bridge CFDictionaryRef)attributes));
+    NSArray *descriptions = [NSArray arrayWithObject:(__bridge id)description.get()];
+
+    auto blockPtr = makeBlockPtr([nsFontName, weakThis = WeakPtr { *this }](CTFontDescriptorMatchingState state, CFDictionaryRef progressParameter) {
+        if (state == kCTFontDescriptorMatchingDidFinish) {
+            RELEASE_LOG(Process, "Font matching finished, progress parameter = %@", (__bridge id)progressParameter);
+            RunLoop::protectedMain()->dispatch([weakThis, nsFontName] {
+                if (RefPtr protectedThis = weakThis.get()) {
+                    auto font = adoptCF(CTFontCreateWithName((__bridge CFStringRef)nsFontName.get(), 0.0, nullptr));
+                    auto cfFontURL = adoptCF(static_cast<CFURLRef>(CTFontCopyAttribute(font.get(), kCTFontURLAttribute)));
+                    URL fontURL(cfFontURL.get());
+                    RELEASE_LOG(Process, "Registering font name %@ with url %s", nsFontName.get(), fontURL.string().utf8().data());
+                    for (auto& process : protectedThis->m_processes)
+                        process->send(Messages::WebProcess::RegisterAdditionalFonts(additionalFonts({ fontURL })), 0);
+                    if (protectedThis->m_additionalFontURLs)
+                        protectedThis->m_additionalFontURLs->append(fontURL);
+                    else
+                        protectedThis->m_additionalFontURLs = { fontURL };
+                }
+            });
+        }
+        return true;
+    });
+
+    if (m_additionalFontURLs) {
```

Comment:
This does not seem right: `WebProcessPool::registerAssetFont` takes `fontName`, which means caller can register a font that's not in `m_additionalFontURLs`. Shouldn't we check if the requested font is in `m_additionalFontURLs` first and return early only if the font already exists?